daemon/mqtt.c

```diff
@@ -0,0 +1,401 @@
+#include "common.h"
+
+struct mosquitto *mosq = NULL;
+
+// Read from the config file -- new section [mqtt]
+// Defaults are supplied
+int mqtt_recv_maximum = 0;      // default 20
+int mqtt_send_maximum = 0;      // default 20
+int mqtt_broker_port = 0;       // default 1883
+char *mqtt_broker_hostname = NULL;  //default localhost
+
+// Set when we have connection up and running from the connection callback
+int mqtt_connection_initialized = 0;
+
+// TODO: integrate with the actual claiming process and get the claiming status from there
+int am_i_claimed()
+{
+    return 1;
+}
+
+
+// This will give the base topic that the agent will publish messages.
+// subtopics will be sent under the base topic e.g.  base_topic/subtopic
+// This is called by mqtt_init(), to compute the base topic once and have
+// it stored internally.
+// Need to check if additional logic should be added to make sure that there
+// is enough information to determine the base topic at init time
+
+
+// TODO: Locking may be needed, depends on the calculation of the base topic and also if we need to switch
+// that on the fly
+
+char *get_publish_base_topic(PUBLISH_TOPIC_ACTION action)
+{
+    static char  *topic = NULL;
+
+    if (unlikely(action == PUBLICH_TOPIC_FREE)) {
+        // TODO: the following is not safe
+        if (likely(topic)) {
+            freez(topic);
+            topic = NULL;
+        }
+        return NULL;
+    }
+
+    if (unlikely(action == PUBLICH_TOPIC_REBUILD)) {
+        get_publish_base_topic(PUBLICH_TOPIC_FREE);
+        return get_publish_base_topic(PUBLICH_TOPIC_GET);
+    }
+
+    if (unlikely(!topic))
+        topic = strdupz("netdata");
+
+    return topic;
+}
+
+
+// Thread cleanup
+static void mqtt_main_cleanup(void *ptr) {
+    struct netdata_static_thread *static_thread = (struct netdata_static_thread *)ptr;
+    static_thread->enabled = NETDATA_MAIN_THREAD_EXITING;
+
+    info("cleaning up...");
+
+    static_thread->enabled = NETDATA_MAIN_THREAD_EXITED;
+}
+
+
+/**
+ * MQTT Main
+ *
+ * This thread will simply call the internal libmosquitto loop to handle
+ * pending requests - both inbound and outbound
+ *
+ * @param ptr is a pointer to the netdata_static_thread structure.
+ *
+ * @return It always returns NULL
+ */
+void *mqtt_main(void *ptr) {
+
+    netdata_thread_cleanup_push(mqtt_main_cleanup, ptr);
+
+    int libmosq_major, libmosq_minor, libmosq_revision, libmosq_version;
+
+    libmosq_version =  mosquitto_lib_version(&libmosq_major, &libmosq_minor, &libmosq_revision);
+
+    info("Detected libmosquitto library version %d, %d.%d.%d",libmosq_version, libmosq_major, libmosq_minor, libmosq_revision);
+
+    while(!netdata_exit) {
+        int rc;
+
+        // TODO: This may change when we have enough info from the claiming itself to avoid wasting 60 seconds
+        // TODO: Handle the unclaim command as well -- we may need to shutdown the connection
+        if (am_i_claimed() == 0) {
+            usleep(60000 * 1000);
```

Comment:
Hi @stelfrag ,

Please change the call from `usleep` to `nanosleep`, because this function was declared obsolete according official documentation http://man7.org/linux/man-pages/man3/usleep.3.html.

Reply:
I will do !! thank you @thiagoftsm

Reply:
Changed to sleep_usec as the rest of the code base in the agent

Reply:
Thank you! :)